README.md
# honeysql-postgres [![Actions Status](https://github.com/nilenso/honeysql-postgres/workflows/CI/badge.svg)](https://github.com/nilenso/honeysql-postgres/actions) [![Clojars Project](https://img.shields.io/clojars/v/nilenso/honeysql-postgres.svg)](https://clojars.org/nilenso/honeysql-postgres) [![cljdoc badge](https://cljdoc.org/badge/nilenso/honeysql-postgres)](https://cljdoc.org/d/nilenso/honeysql-postgres/CURRENT)

PostgreSQL extensions for the widely used [honeysql](https://github.com/jkk/honeysql).

This library aims to extend the features of honeysql to support postgres specific SQL clauses and some basic SQL DDL in addition to the ones supported by the parent library. This keeps honeysql clean and single-purpose, any vendor-specific additions can simply be separate libraries that work on top.

## Honeysql 2.0

Work is underway to enhance the core `honeysql` library to directly provide a large subset of postgres features from this extension. You can read more about it:

- [Note](https://github.com/nilenso/honeysql-postgres/issues/62) from Sean Corfield 
- [Milestone](https://github.com/seancorfield/honeysql/milestone/7) for tracking progress

Regardless of `honeysql` 2.0's feature parity with `honeysql-postgres` and general production stability, this library **will** continue to be maintained going forward.


## Release

### Build
This project follows the version scheme MAJOR.MINOR.COMMITS where MAJOR and MINOR provide some relative indication of the size of the change, but do not follow semantic versioning. In general, all changes endeavor to be non-breaking (by moving to new names rather than by breaking existing names). COMMITS is an ever-increasing counter of commits since the beginning of this repository.

### NPM
From version `0.3.104` and onwards, new releases to [npmjs](https://www.npmjs.com) will not be pushed.
The last available release on npmjs is [0.2.6](https://www.npmjs.com/package/@honeysql/honeysql-postgres).

### Known breaking change
Implementation of `over` has been changed (from `0.2.2`) to accept alias as an option and define the aggregator-function within the over clause and not in the select clause, this allows the inclusion of multiple window-functions which was not possible in the previous implementation.

## Index

- [Usage](#usage)
  - [REPL](#REPL)
  - [distinct on](#distinct-on)
  - [upsert](#upsert)
  - [insert into with alias](#insert-into-with-alias)
  - [over](#over)
  - [create view](#create-view)
  - [create table](#create-table)
  - [drop table](#drop-table)
  - [alter table](#alter-table)
  - [pattern matching](#pattern-matching)
  - [except](#except)
  - [filter](#filter)
  - [within group](#within-group)
  - [SQL functions](#sql-functions)
- [License](#license)

## Usage

### REPL

```clojure
(require '[honeysql.core :as sql]
         '[honeysql.helpers :refer :all :as sqlh]
         '[honeysql-postgres.helpers :as psqlh])
```

### distinct-on

`select` can be written with a `distinct on` clause
``` clojure
(-> (select :column-1 :column-2 :column-3)
    (from :table-name)
    (modifiers :distinct-on :column-1 :column-2)
    (sql/format))
=> ["SELECT DISTINCT ON(column_1, column_2) column_1, column_2, column_3 FROM table_name"]
```

### upsert

`upsert` can be written either way. You can make use of `do-update-set!` over `do-update-set`, if you want to modify the some column values in case of conflicts.
```clojure
(-> (insert-into :distributors)
    (values [{:did 5 :dname "Gizmo Transglobal"}
             {:did 6 :dname "Associated Computing, Inc"}])
    (psqlh/upsert (-> (psqlh/on-conflict :did)
                      (psqlh/do-update-set :dname)))
    (psqlh/returning :*)
    sql/format)
=> ["INSERT INTO distributors (did, dname) VALUES (?, ?), (?, ?) ON CONFLICT (did) DO UPDATE SET dname = EXCLUDED.dname RETURNING *"
    5 "Gizmo Transglobal" 6 "Associated Computing, Inc"]

(-> (insert-into :distributors)
    (values [{:did 23 :dname "Foo Distributors"}])
    (psqlh/on-conflict :did)
    (psqlh/do-update-set! [:dname "EXCLUDED.dname || ' (formerly ' || distributors.dname || ')'"] [:downer "EXCLUDED.downer"])
    sql/format)
=> ["INSERT INTO distributors (did, dname) VALUES (?, ?) ON CONFLICT (did) DO UPDATE SET dname = ?, downer = ?"
    23 "Foo Distributors" "EXCLUDED.dname || ' (formerly ' || distributors.dname || ')'" "EXCLUDED.downer"]
```

### insert into with alias

`insert-into-as` can be used to write insert statements with table name aliased.
```clojure
(-> (psqlh/insert-into-as :distributors :d)
    (values [{:did 5 :dname "Gizmo Transglobal"}
             {:did 6 :dname "Associated Computing, Inc"}])
    (psqlh/upsert (-> (psqlh/on-conflict :did)
                      (psqlh/do-update-set :dname)
                      (where [:<> :d.zipcode "21201"])))
    (psqlh/returning :d.*)
    sql/format)
=> ["INSERT INTO distributors AS d (did, dname) VALUES (?, ?), (?, ?) ON CONFLICT (did) DO UPDATE SET dname = EXCLUDED.dname WHERE d.zipcode <> ? RETURNING d.*"
    5 "Gizmo Transglobal" 6 "Associated Computing, Inc" "21201"]
```

### over

You can make use of `over` to write window functions where it takes in vectors with aggregator functions and window functions along with optional alias like `(over [aggregator-function window-function alias])`, the can be coupled with the `window` clause to write window-function functions with alias that is later defines the window-function, like `(-> (over [aggregator-function :w]) (window :w window-function))`.
```clojure
(-> (select :id)
    (psqlh/over
      [(sql/call :avg :salary) (-> (psqlh/partition-by :department) (order-by [:designation])) :Average]
      [(sql/call :max :salary) :w :MaxSalary])
    (from :employee)
    (psqlh/window :w (psqlh/partition-by :department))
    sql/format)
=> ["SELECT id , avg(salary) OVER (PARTITION BY department ORDER BY designation) AS Average, max(salary) OVER w AS MaxSalary FROM employee WINDOW w AS (PARTITION BY department)"]
```

### create view
`create-view` can be used to create views
```clojure
(-> (psqlh/create-view :metro)
    (select :*)
    (from :cities)
    (where [:= :metroflag "Y"])
    sql/format)
=> ["CREATE VIEW metro AS SELECT * FROM cities WHERE metroflag = ?" "Y"]
```

### create table

`create-table` and `with-columns` can be used to create tables along with the SQL functions, where `create-table` takes a table name as argument and `with-columns` takes a vector of vectors as argument, where the vectors describe the column properties as `[:column-name :datatype :constraints ... ]`.
```clojure
(-> (psqlh/create-table :films)
    (psqlh/with-columns [[:code (sql/call :char 5) (sql/call :constraint :firstkey) (sql/call :primary-key)]
                         [:title (sql/call :varchar 40) (sql/call :not nil)]
                         [:did :integer (sql/call :not nil)]
                         [:date_prod :date]
                         [:kind (sql/call :varchar 10)]])
    sql/format)
=> ["CREATE TABLE films (code char(?) CONSTRAINT firstkey PRIMARY KEY, title varchar(?) NOT NULL, did integer NOT NULL, date_prod date, kind varchar(?))"
    5 40 10]
```

### drop table

`drop-table` is used to drop tables
```clojure
(sql/format (psqlh/drop-table :cities :towns :vilages))
=> ["DROP TABLE cities, towns, vilages"]
```

### alter table

use `alter-table` along with `add-column` & `drop-column` to modify table level details
```clojure
(-> (psqlh/alter-table :employees)
    (psqlh/add-column :address :text)
    sql/format)
=> ["ALTER TABLE employees ADD COLUMN address text"]

(-> (psqlh/alter-table :employees)
    (psqlh/drop-column :address)
    sql/format)
=> ["ALTER TABLE employees DROP COLUMN address"]
```

### create-extension

`create-extension` can be used to create extensions with a given keyword.
```clojure
(-> (psqlh/create-extension :uuid-ossp :if-not-exists? true)
    (sql/format :allow-dashed-names? true
                :quoting :ansi))
=> ["CREATE EXTENSION IF NOT EXISTS \"uuid-ossp\""]
```

### drop-extension
`drop-extension` is used to drop extensions.
```clojure
(-> (psqlh/drop-extension :uuid-ossp)
    (sql/format :allow-dashed-names? true
                :quoting :ansi))
=> ["DROP EXTENSION \"uuid-ossp\""]
```

### pattern matching

The `ilike` and `not-ilike` operators can be used to query data using a pattern matching technique.
- like
```clojure
(-> (select :name)
    (from :products)
    (where [:ilike :name "%name%"])
    sql/format)
=> ["SELECT name FROM products WHERE name ILIKE ?" "%name%"]
```
- not-ilike
```clojure
(-> (select :name)
    (from :products)
    (where [:not-ilike :name "%name%"])
    sql/format)
=> ["SELECT name FROM products WHERE name NOT ILIKE ?" "%name%"]
```

### except

```clojure
(sql/format
  {:except
    [{:select [:ip]}
     {:select [:ip] :from [:ip_location]}]})
=> ["SELECT ip EXCEPT SELECT ip FROM ip_location"]
```
`except-all` works the same way as `except`.

### filter

``` clojure
(-> (select (sql/call :count :*))
    (filter [(sql/call :count :*) (where [:< :i 5]) :foo]
            [(sql/call :count :*) (where [:between :i 3 10]) :bar])
    (from (sql/raw "generate_series(1,10) AS s(i)"))
    (sql/format))
=> ["SELECT count(*) , count(*) FILTER (WHERE i < ?) AS foo, count(*) FILTER (WHERE i BETWEEN ? AND ?) AS bar FROM generate_series(1,10) AS s(i)" 5 3 10]
```

### within group

``` clojure
(-> (select (sql/call :count :*))
    (within-group [(sql/call :percentile_disc (hsql-types/array [0.25 0.5 0.75])) (order-by :s.i) :alias])
    (from (sql/raw "generate_series(1,10) AS s(i)"))
    (sql/format))
=> ["SELECT count(*) , percentile_disc(ARRAY[?, ?, ?]) WITHIN GROUP (ORDER BY s.i) AS alias FROM generate_series(1,10) AS s(i)"
    0.25 0.50 0.75]
```

### SQL functions

The following are the SQL functions added in `honeysql-postgres`
- `not`
```clojure
(sql/format (sql/call :not nil))
=> ["NOT NULL"]
```
- `primary-key`
```clojure
(sql/format (sql/call :primary-key))
=> ["PRIMARY KEY"]

(sql/format (sql/call :primary-key :arg1 :arg2))
=> ["PRIMARY KEY(arg1, arg2)"]
```
- `unique`
```clojure
(sql/format (sql/call :unique))
=> ["UNIQUE"]

(sql/format (sql/call :unique :arg1 :arg2))
=> ["UNIQUE(arg1, arg2)"]
```
- `foreign-key`
```clojure
(sql/format (sql/call :foreign-key))
=> ["FOREIGN KEY"]

(sql/format (sql/call :foreign-key :arg1 :arg2))
=> ["FOREIGN KEY(arg1, arg2)"]
```
- `references`
```clojure
(sql/format (sql/call :references :reftable :refcolumn))
=> ["REFERENCES reftable(refcolumn)"]
```
- `constraint`
```clojure
(sql/format (sql/call :constraint :name))
=> ["CONSTRAINT name"]
```
- `default`
```clojure
(sql/format (sql/call :default :value))
=> ["DEFAULT value"]
```
- `nextval`
```clojure
(sql/format (sql/call :nextval :value))
=> ["nextval('value')"]
```
- `check`
```clojure
(sql/format (sql/call :check [:= :a :b]))
=> ["CHECK(a = b)"]

(sql/format (sql/call :check [:= :a :b] [:= :c :d]))
=> ["CHECK(a = b AND c = d)"]
```

## License

Copyright © 2021 Nilenso

Distributed under the Eclipse Public License, the same as Clojure.


CHANGELOG.md
## 0.4.112
- Upgrade the `honeysql` dependency from `0.9.5` to `1.0.461`
- Support `FILTER` + `WITHIN GROUP` for aggregate expressions
- Comma separate the column names provided to `RETURNING`

## 0.3.104
- Update docs to add syntax for setting multiple [k, v] pairs `ON CONFLICT`
- Add support for `CREATE / DROP EXTENSION`
- Add support for `DISTINCT ON`
- Include `honeysql-posgres.format` directly into `honeysql-postgres.helpers` to avoid leaking the `require` to the consumer
- Stop using lein and shift to deps exclusively
- Add CI and release/publish workflow
- Add support for testing all examples in the README as part of CI

## 0.2.6
- Add ILIKE, NOT ILIKE
- ADD EXCEPT and EXCEPT ALL
- Fix query-values priority

## 0.2.5
- Adds if-exist in drop table

## 0.2.4
- Add if-not-exist option to create-table
- Self-host ClojureScript port
- Add package.json for npm publishing

## 0.2.3
- Allows where clause to be used as an index predicate and update condition during upsert

## 0.2.2-SNAPSHOT
- Breaking change (Revised implementation)
  - over
- Added the following
  - window

## 0.2.1-SNAPSHOT
- Added the following
  - alter-table
  - add-column
  - drop-column

## 0.2.0
- Added the following helpers and corresponding clauses
  - partition by
  - over (window function)
  - create view
  - create table
  - drop table
- Added the following SQL function calls
  - not
  - primary-key
  - foreign-key
  - unique
  - references
  - constraint
  - default
  - nextval
  - check

## 0.1.0
- Added upsert & returning clauses


test/honeysql_postgres/postgres_test.clj
(ns honeysql-postgres.postgres-test
  (:refer-clojure :exclude [update partition-by])
  (:require [honeysql-postgres.format :refer :all]
            [honeysql-postgres.helpers :refer :all]
            [honeysql.helpers :refer :all]
            [honeysql.core :as sql]
            [clojure.test :refer :all]))

(deftest upsert-test
  (testing "upsert sql generation for postgresql"
    (is (= ["INSERT INTO distributors (did, dname) VALUES (5, ?), (6, ?) ON CONFLICT (did) DO UPDATE SET dname = EXCLUDED.dname RETURNING *" "Gizmo Transglobal" "Associated Computing, Inc"]
           (-> (insert-into :distributors)
               (values [{:did 5 :dname "Gizmo Transglobal"}
                        {:did 6 :dname "Associated Computing, Inc"}])
               (upsert (-> (on-conflict :did)
                           (do-update-set :dname)))
               (returning :*)
               sql/format)))
    (is (= ["INSERT INTO distributors (did, dname) VALUES (7, ?) ON CONFLICT (did) DO NOTHING" "Redline GmbH"]
           (-> (insert-into :distributors)
               (values [{:did 7 :dname "Redline GmbH"}])
               (upsert (-> (on-conflict :did)
                           do-nothing))
               sql/format)))
    (is (= ["INSERT INTO distributors (did, dname) VALUES (9, ?) ON CONFLICT ON CONSTRAINT distributors_pkey DO NOTHING" "Antwerp Design"]
           (-> (insert-into :distributors)
               (values [{:did 9 :dname "Antwerp Design"}])
               (upsert (-> (on-conflict-constraint :distributors_pkey)
                           do-nothing))
               sql/format)))
    (is (= ["INSERT INTO distributors (did, dname) VALUES (10, ?), (11, ?) ON CONFLICT (did) DO UPDATE SET dname = EXCLUDED.dname" "Pinp Design" "Foo Bar Works"]
           (sql/format {:insert-into :distributors
                        :values [{:did 10 :dname "Pinp Design"}
                                 {:did 11 :dname "Foo Bar Works"}]
                        :upsert {:on-conflict [:did]
                                 :do-update-set [:dname]}})))
    (is (= ["INSERT INTO distributors (did, dname) VALUES (23, ?) ON CONFLICT (did) DO UPDATE SET dname = ?" "Foo Distributors" "EXCLUDED.dname || ' (formerly ' || d.dname || ')'"]
           (-> (insert-into :distributors)
               (values [{:did 23 :dname "Foo Distributors"}])
               (on-conflict :did)
               (do-update-set! [:dname "EXCLUDED.dname || ' (formerly ' || d.dname || ')'"])
               sql/format)))))


(deftest upsert-where-test
  (is (= ["INSERT INTO user (phone, name) VALUES (?, ?) ON CONFLICT (phone) WHERE phone IS NOT NULL DO UPDATE SET phone = EXCLUDED.phone, name = EXCLUDED.name WHERE user.active = FALSE" "5555555" "John"]
         (sql/format
           {:insert-into :user
            :values      [{:phone "5555555" :name "John"}]
            :upsert      {:on-conflict   [:phone]
                          :where         [:<> :phone nil]
                          :do-update-set {:fields [:phone :name]
                                          :where  [:= :user.active false]}}}))))


(deftest returning-test
  (testing "returning clause in sql generation for postgresql"
    (is (= ["DELETE FROM distributors WHERE did > 10 RETURNING *"]
           (sql/format {:delete-from :distributors
                        :where [:> :did :10]
                        :returning [:*] })))
    (is (= ["UPDATE distributors SET dname = ? WHERE did = 2 RETURNING did dname" "Foo Bar Designs"]
           (-> (update :distributors)
               (sset {:dname "Foo Bar Designs"})
               (where [:= :did :2])
               (returning [:did :dname])
               sql/format)))))

(deftest create-view-test
  (testing "creating a view from a table"
    (is (= ["CREATE VIEW metro AS SELECT * FROM cities WHERE metroflag = ?" "Y"]
           (-> (create-view :metro)
               (select :*)
               (from :cities)
               (where [:= :metroflag "Y"])
               sql/format)))))

(deftest drop-table-test
  (testing "drop table sql generation for a single table"
    (is (= ["DROP TABLE cities"]
           (sql/format (drop-table :cities)))))
  (testing "drop table sql generation for multiple tables"
    (is (= ["DROP TABLE cities, towns, vilages"]
           (sql/format (drop-table :cities :towns :vilages))))))

(deftest create-table-test
  (testing "create table with two columns"
    (is (= ["CREATE TABLE cities (city varchar(80) PRIMARY KEY, location point)"]
           (-> (create-table :cities)
               (with-columns [[:city (sql/call :varchar 80) (sql/call :primary-key)]
                              [:location :point]])
               sql/format))))
  (testing "create table with foreign key reference"
    (is (= ["CREATE TABLE weather (city varchar(80) REFERENCES cities(city), temp_lo int, temp_hi int, prcp real, date date)"]
           (-> (create-table :weather)
               (with-columns [[:city (sql/call :varchar :80) (sql/call :references :cities :city)]
                              [:temp_lo :int]
                              [:temp_hi :int]
                              [:prcp :real]
                              [:date :date]])
               sql/format))))
  (testing "creating table with table level constraint"
    (is (= ["CREATE TABLE films (code char(5), title varchar(40), did integer, date_prod date, kind varchar(10), CONSTRAINT code_title PRIMARY KEY(code, title))"]
           (-> (create-table :films)
               (with-columns [[:code (sql/call :char 5)]
                              [:title (sql/call :varchar 40)]
                              [:did :integer]
                              [:date_prod :date]
                              [:kind (sql/call :varchar 10)]
                              [(sql/call :constraint :code_title) (sql/call :primary-key :code :title)]])
               sql/format))))
  (testing "creating table with column level constraint"
    (is (= ["CREATE TABLE films (code char(5) CONSTRAINT firstkey PRIMARY KEY, title varchar(40) NOT NULL, did integer NOT NULL, date_prod date, kind varchar(10))"]
           (-> (create-table :films)
               (with-columns [[:code (sql/call :char 5) (sql/call :constraint :firstkey) (sql/call :primary-key)]
                              [:title (sql/call :varchar 40) (sql/call :not nil)]
                              [:did :integer (sql/call :not nil)]
                              [:date_prod :date]
                              [:kind (sql/call :varchar 10)]])
               sql/format))))
  (testing "creating table with columns with default values"
    (is (= ["CREATE TABLE distributors (did integer PRIMARY KEY DEFAULT nextval('serial'), name varchar(40) NOT NULL)"]
           (-> (create-table :distributors)
               (with-columns [[:did :integer (sql/call :primary-key) (sql/call :default (sql/call :nextval :serial))]
                              [:name (sql/call :varchar 40) (sql/call :not nil)]])
               sql/format))))
  (testing "creating table with column checks"
    (is (= ["CREATE TABLE products (product_no integer, name text, price numeric CHECK(price > 0), discounted_price numeric, CHECK(discounted_price > 0 AND price > discounted_price))"]
           (-> (create-table :products)
               (with-columns [[:product_no :integer]
                              [:name :text]
                              [:price :numeric (sql/call :check [:> :price 0])]
                              [:discounted_price :numeric]
                              [(sql/call :check [:> :discounted_price 0] [:> :price :discounted_price])]])
               sql/format)))))

(deftest over-test
  (testing "window function over on select statemt"
    (is (= ["SELECT id , avg(salary) OVER (PARTITION BY department ORDER BY designation) AS Average, max(salary) OVER w AS MaxSalary FROM employee WINDOW w AS (PARTITION BY department)"]
           (-> (select :id)
               (over
                [(sql/call :avg :salary) (-> (partition-by :department) (order-by [:designation])) :Average]
                [(sql/call :max :salary) :w :MaxSalary])
               (from :employee)
               (window :w (partition-by :department))
               sql/format)))))

(deftest alter-table-test
  (testing "alter table add column generates the required sql"
    (is (= ["ALTER TABLE employees ADD COLUMN address text"]
           (-> (alter-table :employees)
               (add-column :address :text)
               sql/format))))
  (testing "alter table drop column generates the required sql"
    (is (= ["ALTER TABLE employees DROP COLUMN address"]
           (-> (alter-table :employees)
               (drop-column :address)
               sql/format))))
  (testing "alter table rename column generates the requred sql"
    (is (= ["ALTER TABLE employees RENAME COLUMN address TO homeaddress"]
           (-> (alter-table :employees)
               (rename-column :address :homeaddress)
               sql/format))))
  (testing "alter table rename table generates the required sql"
    (is (= ["ALTER TABLE employees RENAME TO managers"]
           (-> (alter-table :employees)
               (rename-table :managers)
               sql/format)))))

(deftest insert-into-with-alias
  (testing "insert into with alias"
    (is (= ["INSERT INTO distributors AS d (did, dname) VALUES (5, ?), (6, ?) ON CONFLICT (did) DO UPDATE SET dname = EXCLUDED.dname WHERE d.zipcode <> ? RETURNING d.*" "Gizmo Transglobal" "Associated Computing, Inc" "21201"]
           (-> (insert-into-as :distributors :d)
               (values [{:did 5 :dname "Gizmo Transglobal"}
                        {:did 6 :dname "Associated Computing, Inc"}])
               (upsert (-> (on-conflict :did)
                           (do-update-set :dname)
                           (where [:<> :d.zipcode "21201"])))
               (returning :d.*)
               sql/format)))))


project.clj
(defproject nilenso/honeysql-postgres "0.2.6"
  :description "PostgreSQL extension for honeysql"
  :url "https://github.com/nilenso/honeysql-postgres"
  :license {:name "Eclipse Public License"
            :url  "http://www.eclipse.org/legal/epl-v10.html"}
  :dependencies [[org.clojure/clojure "1.10.1"]
                 [honeysql "1.0.444"]]
  :deploy-repositories [["clojars" {:url      "https://clojars.org/nilenso/honeysql-postgres"
                                    :username :env/CLOJARS_USERNANE
                                    :password :env/CLOJARS_DEPLOY_TOKEN}]]
  :plugins [[lein-cljfmt "0.7.0"]]
  :tach {:test-runner-ns 'honeysql-postgres.postgres-test
         :source-paths   ["src" "test"]}
  :profiles {:dev {:plugins [[lein-tach "0.4.0"]]}
             :ci  {:dependencies [[seancorfield/readme "1.0.16"]]
                   :aliases      {"verify-readme" ["run" "-m" "seancorfield.readme"]}}})


src/honeysql_postgres/util.clj

(ns ^{:doc "Postgres Honeysql utils"}
    honeysql-postgres.util
  (:refer-clojure :exclude [format partition-by])
  (:require [honeysql.format :refer :all]))

(defn get-first
  "Returns the first element if the passed argument is a collection, else return the passed argument
   as is."
  [x]
  (if (sequential? x)
    (first x)
    x))

(defn comma-join-args
  "Returns the args comma-joined after applying to-sql to them"
  [args]
  (if (nil? args)
    ""
    (->> args
         (map to-sql)
         comma-join
         paren-wrap)))

(defn prep-check
  "Adds a logical `:and` operation if args has multiple vectors
   Eg - (prep-check '([:= :a :b])) => [:= :a :b]
        (prep-check '([:= :a :b] [:< :a :c])) => [:and [:= :a :b] [:< :a :c]]"
  [args]
  (let [preds (if (= 1 (count args))
                  (first args)
                  args)
        [logic-op preds] (if (keyword? (first preds))
                           [(first preds) (rest preds)]
                           [:and preds])]
    (if (= 1 (count preds))
      (first preds)
      (into [logic-op] preds))))


src/honeysql_postgres/helpers.clj
(ns honeysql-postgres.helpers
  (:refer-clojure :exclude [update partition-by])
  (:require [honeysql.helpers :refer :all]))

;; Extension of the honeysql helper funcitons for postgreSQL

(defn do-nothing [m]
  "Accepts a map and append {:do-nothing []} to it"
  (assoc m :do-nothing []))

(defhelper do-update-set [m args]
  (assoc m :do-update-set (collify args)))

(defhelper do-update-set! [m args]
  (assoc m :do-update-set! args))

(defhelper on-conflict [m args]
  (assoc m :on-conflict args))

(defhelper on-conflict-constraint [m args]
  (assoc m :on-conflict-constraint args))

(defhelper upsert [m args]
  (if (plain-map? args)
    (assoc m :upsert args)
    (assoc m :upsert (first args))))

(defhelper returning [m fields]
  (assoc m :returning (collify fields)))

(defhelper create-view [m viewname]
  (assoc m :create-view (collify viewname)))

(defhelper create-table [m tablename]
  (assoc m :create-table (collify tablename)))

(defhelper with-columns [m args]
  (assoc m :with-columns args))

(defhelper drop-table [m tablenames]
  (assoc m :drop-table (collify tablenames)))

(defhelper over [m args]
  (assoc m :over (collify args)))

(defhelper window [m args]
  (assoc m :window args))

(defhelper partition-by [m fields]
  (assoc m :partition-by (collify fields)))

(defhelper alter-table [m fields]
  (assoc m :alter-table (collify fields)))

(defhelper add-column [m fields]
  (assoc m :add-column (collify fields)))

(defhelper drop-column [m fields]
  (assoc m :drop-column (collify fields)))

(defhelper rename-column [m fields]
  (assoc m :rename-column fields))

(defhelper rename-table [m fields]
  (assoc m :rename-table (collify fields)))

(defhelper insert-into-as [m fields]
  (assoc m :insert-into-as (collify fields)))


src/honeysql_postgres/format.clj

(ns ^{:doc "Extension of the honeysql format functions specifically for postgreSQL"}
    honeysql-postgres.format
  (:refer-clojure :exclude [format partition-by])
  (:require [honeysql.format :refer :all]
            [honeysql-postgres.util :refer [get-first comma-join-args prep-check]]))

(def ^:private custom-additions
  {:create-table 10
   :drop-table 10
   :alter-table 20
   :add-column 30
   :drop-column 40
   :create-view 40
   :insert-into-as 60
   :over 55
   :partition-by 165
   :window 195
   :upsert 225
   :on-conflict 230
   :on-conflict-constraint 230
   :do-update-set 235
   :do-update-set! 235
   :do-nothing 235
   :returning 240})

(def ^:private postgres-clause-priorities
  "Determines the order that clauses will be placed within generated SQL"
  (merge {:with 30
          :with-recursive 40
          :select 50
          :insert-into 60
          :update 70
          :delete-from 80
          :columns 90
          :set 100
          :from 110
          :join 120
          :left-join 130
          :right-join 140
          :full-join 150
          :where 160
          :group-by 170
          :having 180
          :order-by 190
          :limit 200
          :offset 210
          :lock 215
          :values 220
          :query-values 250}
         custom-additions))

(defn override-default-clause-priority
  "Override the default cluse priority set by honeysql"
  []
  (doseq [[k v] postgres-clause-priorities]
    (register-clause! k v)))

(defmethod fn-handler "not" [_ value]
  (str "NOT " (to-sql value)))

(defmethod fn-handler "primary-key" [_ & args]
  (str "PRIMARY KEY" (comma-join-args args)))

(defmethod fn-handler "unique" [_ & args]
  (str "UNIQUE" (comma-join-args args)))

(defmethod fn-handler "foreign-key" [_ & args]
  (str "FOREIGN KEY" (comma-join-args args)))

(defmethod fn-handler "references" [_ reftable refcolumn]
  (str "REFERENCES " (to-sql reftable) (paren-wrap (to-sql refcolumn))))

(defmethod fn-handler "constraint" [_ name]
  (str "CONSTRAINT " (to-sql name)))

(defmethod fn-handler "default" [_ value]
  (str "DEFAULT " (to-sql value)))

(defmethod fn-handler "primary-key" [_ & args]
  (str "PRIMARY KEY" (util/comma-join-args args)))

(defmethod fn-handler "nextval" [_ value]
  (str "nextval('" (to-sql value) "')"))

(defmethod fn-handler "check" [_ & args]
  (let [preds (format-predicate* (prep-check args))
        pred-string (if (= 1 (count args))
                      (paren-wrap preds)
                      preds)]
    (str "CHECK" pred-string)))

;; format-clause multimethods used to format various sql clauses as defined

(defmethod format-clause :on-conflict-constraint [[_ k] _]
  (str "ON CONFLICT ON CONSTRAINT " (-> k
                                        get-first
                                        to-sql)))

(defmethod format-clause :on-conflict [[_ ids] _]
  (str "ON CONFLICT " (comma-join-args ids)))

(defmethod format-clause :do-nothing [_ _]
  "DO NOTHING")

(defmethod format-clause :do-update-set! [[_ values] _]
  (str "DO UPDATE SET " (comma-join (for [[k v] values]
                                      (str (to-sql k) " = " (to-sql v))))))

(defmethod format-clause :do-update-set [[_ values] _]
  (let [fields (or (:fields values) values)
        where  (:where values)]
    (str "DO UPDATE SET "
      (comma-join (map #(str (to-sql %) " = EXCLUDED." (to-sql %)) fields))
      (when where
        (str " WHERE " (format-predicate* where))))))

(defn- format-upsert-clause [upsert]
  (let [ks (keys upsert)]
    (map #(format-clause % (find upsert %)) upsert)))

(defmethod format-clause :upsert [[_ upsert] _]
  (space-join (format-upsert-clause upsert)))

(defmethod format-clause :returning [[_ fields] _]
  (str "RETURNING " (comma-join (map to-sql fields))))

(defmethod format-clause :create-view [[_ viewname] _]
  (str "CREATE VIEW " (-> viewname
                          get-first
                          to-sql) " AS"))

(defmethod format-clause :create-table [[_ tablename] _]
  (str "CREATE TABLE " (-> tablename
                           get-first
                           to-sql)))

(defmethod format-clause :with-columns [[_ columns] _]
  (paren-wrap (->> columns
                   get-first
                   (map #(space-join (map to-sql %)))
                   comma-join)))

(defmethod format-clause :drop-table [[_ tables] _]
  (str "DROP TABLE " (->> tables
                          (map to-sql)
                          comma-join)))

(defn- format-over-clause [exp]
  (str
   (-> exp first to-sql)
   " OVER "
   (-> exp second to-sql)
   (when-let [alias (-> exp rest second)]
     (str " AS " (to-sql alias)))))

(defmethod format-clause :over [[_ fields] complete-sql-map]
  (str
   ;; if the select clause has any columns in it then add a comma before the
   ;; window functions
   (if (seq (:select complete-sql-map)) ", ")
   (->> fields
        (map format-over-clause)
        comma-join)))

(defmethod format-clause :window [[_ [window-name fields]] _]
  (str "WINDOW " (to-sql window-name) " AS " (to-sql fields)))

(defmethod format-clause :partition-by [[_ fields] _]
  (str "PARTITION BY " (->> fields
                            (map to-sql)
                            comma-join)))

(defmethod format-clause :alter-table [[_ tablename] _]
  (str "ALTER TABLE " (-> tablename
                          get-first
                          to-sql)))

(defmethod format-clause :add-column [[_ fields] _]
  (str "ADD COLUMN " (->> fields
                         (map to-sql)
                         space-join)))

(defmethod format-clause :drop-column [[_ fields] _]
  (str "DROP COLUMN " (->> fields
                           get-first
                           to-sql)))

(defmethod format-clause :rename-column [[_ [oldname newname]] _]
  (str "RENAME COLUMN " (to-sql oldname) " TO " (to-sql newname)))

(defmethod format-clause :rename-table [[_ newname] _]
  (str "RENAME TO " (-> newname
                        get-first
                        to-sql)))

(defmethod format-clause :insert-into-as [[_ [table-name table-alias]] _]
  (str  "INSERT INTO " (to-sql table-name) " AS " (to-sql table-alias)))

(override-default-clause-priority)
